### ISSUE TITLE
[FR] decouple UI thread from file preview

### ISSUE THREAD

**Issue by laur89** [vifm/vifm#481]
Currently scrolling in directory with file preview enabled in 2nd pane could get extremely sluggish if the preview command isn't snappy enough; eg having following config to use [bat](https://github.com/sharkdp/bat) for previews causes scrolling in a directory containing many shell scripts to become noticeably slow:

    " remaining files:
    fileviewer * bat --color always --plain %c
    fileviewer .* bat --color always --plain %c

Would it be feasible to separate the UI thread from one dealing with previews?

**Comment by xaizek** [comment]
Well, adding a thread for that might not be necessary at all. You already have independent processes. It's displaying which needs to be made asynchronous and not wait for data.

I did keep this in mind (there were similar requests before) while refactoring to add file previews to miller view. But, of course, it now needs more work to make it work asynchronously and manage processes (e.g., kill old ones when you switch files).

**Comment by patroclos** [comment]
I'm currently taking a look at the code around quickviews, since I have a similar issue when using [vifmimg](https://github.com/cirala/vifmimg), which does pdf and image previews that can be fairly heavy.

However, even small images make scrolling sluggish and holding 'j' for a couple of seconds, then releasing it causes it to continue scrolling for some time.

I found out, in that case the problem comes from the usleep(50000) inside view_file for non-textual viewers, such as image previews.

Maybe the time to sleep should be made configurable, so you can turn it off or reduce it if you don't need it.

**Comment by patroclos** [comment]
@xaizek I probably would implement it like the following, can you tell me if it doesn't make sense because of something I don't know?

I think for graphical viewers, it's not sufficient to just kill the viewer, since that might leave stuff like ueberzug in a weird halfway-done state.

But then again, I don't know if running every viewer from start to finish is advisable, when scrolling through a list of heavy files fast.

Maybe the best solution to this is to send a SIGINT to the viewer, if it's still running when the cleanup is requested, and wait until the job isn't running until executing the cleanup command.

So in essence:

* create background jobs when running the viewer and cleanup commands
* keep track of all running view and cleanup jobs
* on cleanup
  * graphical viewer
    * if the initial view job is still running, send SIGINT to it and defer the cleanup until it's dead
    * otherwise, just run the cleanup
  * non-graphical viewer
    * if the viewer job is running, just kill it and run the cleanup

In addition to that, it would be good to track and treat the deferred stuff as a queue,
so we can check the state of that queue when we scroll and trigger new previews.

This would allow us to collapse:
* Cleanup A
* Show B
* Cleanup B
* Show C <- the new one

into

* Cleanup A
* Show C

**Comment by xaizek** [comment]
I wasn't really thinking about graphics preview while writing my response, they do make things more complicated. Your analysis regarding `SIGINT` and queues makes sense.

> I found out, in that case the problem comes from the usleep(50000) inside view_file for non-textual viewers, such as image previews.

Right, this isn't desirable for Ueberzug, maybe something like `'previewoptions'` could be added to configure the delay.

**Comment by xaizek** [comment]
@patroclos, added `'previewoptions'`, thanks for the suggestion. With this
```vim
:set previewoptions+=graphicsdelay:0
```
previewing with `w3m` lags much less and I was blaming it on `w3m`.

**Comment by laur89** [comment]
@xaizek would it be reasonable to make value of `0` default?

**Comment by xaizek** [comment]
> would it be reasonable to make value of `0` default?

With the delay it works in more terminals. Also if default value of a new option won't match old behaviour, it might look like something got broken. So I'm not sure it's a good idea.

**Comment by xaizek** [comment]
The main part has just landed on `master`. Graphical previewers (doesn't include Sixel and other pass-through viewers) is what's left as they are still synchronous. Need to think how to do the clearing and queuing for them properly and what effects it might have.

So don't do catch-all things like
```vim
fileviewer * universal-viewer %c %px %py %pw %ph %pc clear
```
or everything will be synchronous. Separate graphical and non-graphical viewers.

**Comment by seebye** [comment]
> Graphical previewers (doesn't include Sixel and other pass-through viewers)

@xaizek Ueberzug works flawlessly with `set previewoptions+=graphicsdelay:0`. Actually that's how ueberzug is expected to be used.

**Comment by xaizek** [comment]
Yes, Ueberzug previewer just sends a request to do the preview for pictures, but there are other things like:
```bash
convert -coalesce "$PWD/$6" "/tmp$PWD/$6/$6.png"
fontpreview -i "$6" -o "/tmp${PWD}/$6.png"
ffmpeg -i "$6" "/tmp${PWD}/$6.png" -y &> /dev/null
epub-thumbnailer "$6" "/tmp$PWD/$6.png" 1024
ffmpegthumbnailer -i "$PWD/$6" -o "/tmp$PWD/$6.png" -s 0 -q 10
```
which can add a delay, so even for Ueberzug it might be relevant to a degree. Worth thinking about at least.

**Comment by seebye** [comment]
As I initially wrote the ueberzug script someone made the effort to animate gifs.  
I thought it was a funny idea, but it had a problem as it blocked vifm.  
I looked a bit around in your source code and noticed that you're waiting till the file descriptors are closed.  
So I made an asynchronous version by closing the file descriptors at the beginning of the bash script.  
```bash
exec 1>&-
exec 2>&-
exec 0<&-
```
It worked well, so maybe that's a solution to make it also asynchronous?
